lib/Sema/CSBindings.cpp

```diff
@@ -204,14 +204,26 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
 
     case ConstraintKind::BridgingConversion:
     case ConstraintKind::CheckedCast:
-    case ConstraintKind::DynamicTypeOf:
     case ConstraintKind::EscapableFunctionOf:
     case ConstraintKind::OpenedExistentialOf:
     case ConstraintKind::KeyPath:
     case ConstraintKind::KeyPathApplication:
       // Constraints from which we can't do anything.
       continue;
 
+    case ConstraintKind::DynamicTypeOf: {
+      // Direct binding of the left-hand side could result
+      // in `DynamicTypeOf` failure if right-hand side is
+      // bound (because 'Bind' requires equal types to
+      // succeed), or left is bound to Any which is not an
+      // [existential] metatype.
+      if (constraint->getFirstType()->isEqual(typeVar))
+        return result;
```

Comment:
It's a silly example, but for the following we're currently able to disambiguate based on the constraint from the parameter type:

```
class C {
   typealias T = Int
}

class D : C {
   typealias T = Float
}

func foo(_: Int.Type.Type) {}
foo(type(of: D.T.self))
```